connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/connect/client/SparkConnectClient.scala

```diff
@@ -60,14 +64,36 @@ private[sql] class SparkConnectClient(
     new ArtifactManager(userContext, sessionId, channel)
   }
 
+  private val retryParameters: SparkConnectClient.RetryParameters = configuration.retryParameters
+
+  @tailrec private[client] final def retry[T](fn: => T, retries: Int = 0): T = {
+    if (retries > retryParameters.max_retries) {
+      throw new IllegalArgumentException(s"retries must not exceed retryParameters.max_retries")
+    }
+    Try {
```

Comment:
The way of using `Try` seems slightly odd. I think you should either:

```scala
Try(fn).getOrElse(...)
```

or just plain try catch:

```scala
    try {
      fn
    } catch {
      case NonFatal(e) if
          retryParameters.should_retry(e) && retries < retryParameters.max_retries =>
        Thread.sleep(
          (retryParameters.max_backoff min retryParameters.initial_backoff * Math
            .pow(retryParameters.backoff_multiplier, retries)).toMillis)
        retry(fn, retries + 1)
    }
```

Reply:
I did it this way to be able to use the @tailrec optimizations, but I now found a way to do it with plain old try catch.